toolkit/tools/pkg/imagecustomizerlib/imageutils.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"path/filepath"
+	"sort"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/configuration"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/installutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+)
+
+type installOSFunc func(imageChroot *safechroot.Chroot) error
+
+func connectToExistingImage(imageFilePath string, buildDir string, chrootDirName string) (*ImageConnection, error) {
+	imageConnection := NewImageConnection()
+
+	err := connectToExistingImageHelper(imageConnection, imageFilePath, buildDir, chrootDirName)
+	if err != nil {
+		imageConnection.Close()
+		return nil, err
+	}
+
+	return imageConnection, nil
+
+}
+
+func connectToExistingImageHelper(imageConnection *ImageConnection, imageFilePath string,
+	buildDir string, chrootDirName string,
+) error {
+	// Connect to image file using loopback device.
+	err := imageConnection.ConnectLoopback(imageFilePath)
+	if err != nil {
+		return err
+	}
+
+	// Look for all the partitions on the image.
+	newMountDirectories, mountPoints, err := findPartitions(buildDir, imageConnection.Loopback().DevicePath())
+	if err != nil {
+		return fmt.Errorf("failed to find disk partitions:\n%w", err)
+	}
+
+	// Create chroot environment.
+	imageChrootDir := filepath.Join(buildDir, chrootDirName)
+
+	err = imageConnection.ConnectChroot(imageChrootDir, false, newMountDirectories, mountPoints)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createNewImage(filename string, diskConfig imagecustomizerapi.Disk,
+	partitionSettings []imagecustomizerapi.PartitionSetting, bootType imagecustomizerapi.BootType, buildDir string,
+	chrootDirName string, installOS installOSFunc,
+) (*ImageConnection, error) {
+	imageConnection := &ImageConnection{}
+
+	err := createNewImageHelper(imageConnection, filename, diskConfig, partitionSettings, bootType, buildDir,
+		chrootDirName, installOS,
+	)
+	if err != nil {
+		imageConnection.Close()
```

Comment:
why cleanclose is not needed here?

Reply:
`CleanClose()` is used when you want to make sure the resource fully closes before continuing.

`Close()` is used when you just want the resource to close, and you don't care if it succeeds or is done asynchronously. This is typically used when an error occurs.